design/resource-management.md

```diff
@@ -0,0 +1,329 @@
+in inconsistent behavior. This document descibes the state of 1.6, and a suggested implementation
+for 1.7 version of Kata.
+
+Before diving into the gaps and behavior exhibited in Kata Containers, it is important to have
+a thorough understanding of how cgroups are leveraged by Kubernetes.  It is both straight forward
+and confusing.  An in-depth guide is available for background in [mcastelino's gist](https://gist.github.com/mcastelino/b8ce9a70b00ee56036dadd70ded53e9f).
+This may be good content to include in our repository eventually, or as part of the Kata blog series,
+but let's leave it out of the scope of this initial documentation.
+
+This document is part of the `cgroup-sprint` GitHub milestone, and can be observed in the milestone's
+[GitHub project](https://github.com/orgs/kata-containers/projects/17).
+
+
+## Existing Behavior (Kata 1.6):
+
+### Behavior Observed using various upper layer tools
+
+To exhibit current behavior, we utilize a simple guaranteed pod description:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: guar-runc
+spec:
+  containers:
+  - name: cont-2cpu-400m
+    image: busybox
+    resources:
+      limits:sadf
+        cpu: 2
+        memory: "400Mi"
+    command: ["md5sum"]
+    args: ["/dev/urandom"]
+  - name: cont-3cpu-200m
+    image: busybox
+    resources:
+      limits:
+        cpu: 3
+        memory: "200Mi"
+    command: ["md5sum"]
+    args: ["/dev/urandom"]
+```
+We'll show the behavior starting with the simplest scenario, Docker, followed by containerd
+and CRI-O.
+
+#### In Docker
+
+#### In Kubernetes + Containerd
+
+For each container in the pod, a cgroup is created within the pod cgroup (ie, under `/sys/fs/cgroup/*/kubepod/pod.*/`
+for a guaranteed pod). This is not necessary; only a single cgroup which constrains the hypervisor
+appropriately is required.
+
+##### Where are all the vCPUS
+
+The vCPUs are placed under the pause container:
+
+```bash
+# for i in `ls pod*/**/tasks`; do echo $i && for j in `cat $i`; do pstree -pt $j;done; done;
+podf277e232-5ca6-11e9-b514-000d3a6d0876/2692eaedb55f8cfd1b9aadcbc5e3f0ac527cb39ff26d31877f1be5a495b966c1/tasks
+podf277e232-5ca6-11e9-b514-000d3a6d0876/6689f72eef2161b85d5d57cb9f4670ae4e08f551d9aeb4b28efb67eb306034d8/tasks
+podf277e232-5ca6-11e9-b514-000d3a6d0876/9d17d1c5d075ca42dfefb58ef7b82c8b1b234cc128ebf9332b902b866c0ebed7/tasks
+{qemu-system-x86}(24011)
+{qemu-system-x86}(24093)
+{qemu-system-x86}(24097)
+{qemu-system-x86}(24156)
+{qemu-system-x86}(24157)
+{qemu-system-x86}(24158)
+```
+
+In this case, `9d17d1` is the pause, as you can see based on the summation for `cpu.cfs_quota_us`:
+```bash
+# cat */cpu.cfs_quota_us
+300000
+200000
+500000
+```
+One drawback of this is that it assumes the existence of a pause container. This is an assumption
+based on the implementation of containerd/cri-o.  On the plus side, the cgroup is placed directly under
+the pod cgroup, which is created and managed by Kubelet.  Overall, this isn't terrible.
+
```

Comment:
However CPU resources are proportional.  Hence if any of the sibling cgroups consume resources the summation logic does not work.